packages/runtime-core/__tests__/rendererOptimizedMode.spec.ts

```diff
@@ -1,3 +1,401 @@
+    renderWithBlock(() => [
+      createVNode('p', { id: 'foo' }, '', PatchFlags.PROPS, ['id'])
+    ])
+
+    expect(inner(root)).toBe('<div><p id="foo"></p></div>')
+    expect(block!.dynamicChildren!.length).toBe(1)
+    expect(serialize(block!.dynamicChildren![0].el as TestElement)).toBe(
+      '<p id="foo"></p>'
+    )
+
+    renderWithBlock(() => [
+      createVNode('p', { id: 'bar' }, '', PatchFlags.PROPS, ['id'])
+    ])
+
+    expect(inner(root)).toBe('<div><p id="bar"></p></div>')
+    expect(block!.dynamicChildren!.length).toBe(1)
+    expect(serialize(block!.dynamicChildren![0].el as TestElement)).toBe(
+      '<p id="bar"></p>'
+    )
+  })
+
+  test('PatchFlags: PatchFlags.FULL_PROPS', async () => {
+    let propName = 'foo'
+
+    renderWithBlock(() => [
+      createVNode('p', { [propName]: 'dynamic' }, '', PatchFlags.FULL_PROPS)
+    ])
+
+    expect(inner(root)).toBe('<div><p foo="dynamic"></p></div>')
+    expect(block!.dynamicChildren!.length).toBe(1)
+    expect(serialize(block!.dynamicChildren![0].el as TestElement)).toBe(
+      '<p foo="dynamic"></p>'
+    )
+
+    propName = 'bar'
+    renderWithBlock(() => [
+      createVNode('p', { [propName]: 'dynamic' }, '', PatchFlags.FULL_PROPS)
+    ])
+
+    expect(inner(root)).toBe('<div><p bar="dynamic"></p></div>')
+    expect(block!.dynamicChildren!.length).toBe(1)
+    expect(serialize(block!.dynamicChildren![0].el as TestElement)).toBe(
+      '<p bar="dynamic"></p>'
+    )
+  })
+
+  // the order and length of the list will not change
+  test('PatchFlags: PatchFlags.STABLE_FRAGMENT', async () => {
+    let list = ['foo', 'bar']
+    render(
+      (openBlock(),
+      (block = createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode('p', null, item, PatchFlags.TEXT)
+        }),
+        PatchFlags.STABLE_FRAGMENT
+      ))),
+      root
+    )
+
+    expect(inner(root)).toBe('<p>foo</p><p>bar</p>')
+    expect(block.dynamicChildren!.length).toBe(2)
+    expect(serialize(block.dynamicChildren![0].el as TestElement)).toBe(
+      '<p>foo</p>'
+    )
+    expect(serialize(block.dynamicChildren![1].el as TestElement)).toBe(
+      '<p>bar</p>'
+    )
+
+    list = list.map(item => item.repeat(2))
+    render(
+      (openBlock(),
+      createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode('p', null, item, PatchFlags.TEXT)
+        }),
+        PatchFlags.STABLE_FRAGMENT
+      )),
+      root
+    )
+
+    expect(inner(root)).toBe('<p>foofoo</p><p>barbar</p>')
+    expect(block.dynamicChildren!.length).toBe(2)
+    expect(serialize(block.dynamicChildren![0].el as TestElement)).toBe(
+      '<p>foofoo</p>'
+    )
+    expect(serialize(block.dynamicChildren![1].el as TestElement)).toBe(
+      '<p>barbar</p>'
+    )
+  })
+
+  // Fragment as a block, since the `UNKEYED_FRAGMENT` flag will cause bail out of the optimization mode,
+  // so we disable tracking here.
```

Comment:
It's possible for a Fragment with `UNKEYED_FRAGMENT` flag to be in tracking mode, e.g. a template with multiple root nodes, or child fragments of `<template v-for>`.

Reply:
Hey Evan, thank you for your teaching, but I still have some doubts: Isn't a template with multiple root nodes a stable fragment? I tried to find the use case of a Fragment with `UNKEYED_FRAGMENT` flag in tracking mode, but I failed.

Reply:
Yeah they should be stable fragments. I got confused myself 😅